### PR TITLE
add Custom Colormap querystring option

### LINKED ISSUE #153: User specified "custom" colormap
Is there a way to allow the client to specify a custom colormap "dynamically"?  For example, by providing a series of pixel+color pairs and dynamically generating a linear gradient colormap on the fly during rendering?

If such a capability doesn't exist, would you have any advice on how it might be best implemented?

### PR DESCRIPTION
**Pull request by vincentsarago**
closes #153  

This PR does:
- rename `color_map` query parameter to `colormap_name` 
- add `colormap` option.

The `colormap` option accepts JSON encoded object in form of:

```
{
    "1": {RGB(A)}
}
```

the `{RBG(A)}` value can either be a list of R,G,B (and Alpha) values or an hex color representation

cc @kylebarron @blazetopher

### CODE REVIEW

File: titiler/dependencies.py [hunk 1]
```diff
@@ -68,12 +69,67 @@ def TMSParams(
     return tms.get(TileMatrixSetId.name)
 
 
+def parse_color(value: Union[List[int], str]) -> List[int]:
```
Comment (vincentsarago) [inline titiler/dependencies.py:72]: I think I'm going to move this to rio-tiler

File: titiler/dependencies.py [hunk 2]
```diff
@@ -68,12 +69,67 @@ def TMSParams(
...
+    if len(value) == 3:
+        value += [255]
+
+    return value
+
+
 def ColorMapParams(
-    color_map: ColorMapName = Query(None, description="Colormap name",)
+    colormap_name: ColorMapName = Query(None, description="Colormap name"),
+    colormap: str = Query(None, description="JSON encoded custom Colormap"),
```
Comment (vincentsarago) [inline titiler/dependencies.py:121]: **breaking change**

File: titiler/dependencies.py [hunk 3]
```diff
@@ -68,12 +69,67 @@ def TMSParams(
...
 ) -> Optional[Dict]:
     """Colormap Dependency."""
-    if color_map:
-        return cmap.get(color_map.value)
+    if colormap_name:
+        return cmap.get(colormap_name.value)
+
+    if colormap:
+        return json.loads(
```
Comment (vincentsarago) [inline titiler/dependencies.py:128]: should we also accept base64 encoded string?

File: titiler/dependencies.py (hunk 1 above)
Comment (vincentsarago) [reply titiler/dependencies.py:72]: https://github.com/cogeotiff/rio-tiler/pull/361

File: titiler/dependencies.py (hunk 3 above)
Comment (geospatial-jeff) [reply titiler/dependencies.py:128]: I think base64 is less weird than JSON for get requests

Comment (vincentsarago) [reply titiler/dependencies.py:128]: another way maybe would be to pass a list of value like 
```
colormap=1,#FFFF00&colormap=2,#FF00FF

#equivalent of

colormap={"1": "#FFFF00", "2":"#FF00FF"}
```

**Comment by vincentsarago** [comment]
following some discussion on slack, 

Right now a user could define a **discrete** colormap by passing a JSON encoded object in form of 
```
colormap={"1": "#FFFF00", "2":"#FF00FF"}
``` 

I understand that this is quite uncommon for GET request to receive `dict` in a query parameter, but we don't really have a choice (because we can't store a colormap `somewhere`. 

we need a way to pass a `list` of key/value, and for now this PR uses a JSON dict.

`List` seems to be an accepted type for query parameter, using the form `opt=1&opt=2` (FastAPI will resolve this to `opt=[1, 2]`. So maybe a form like `colormap=1,#FFFF00&colormap=2,#FF00FF` would be better 🤷 


## Options 

- JSON encoded object {note: would need to be decoded and re-encoded in tilejson endpoint)

`...preview.png?colormap={"1": "#FFFF00", "2":"#FF00FF"}`  (should be url encoded)

- base64 encoded dict (seems to be a nice option but pretty custom)

`...preview.png?colormap=dsaidhauhdasiojhd928e1uqwijs09qus0x8ahxas`

- List defined colormap value (restricted to hex color definition and custom key/color `1,#fff` form)

`...preview.png?colormap=1,#FFFF00&colormap=2,#FF00FF`

👂

**Comment by vincentsarago** [comment]
Note: the `#` shouldn't be used in a non-encoded url

**Comment by vincentsarago** [comment]
#### Option 1
JSON encoded object {note: would need to be decoded and re-encoded in tilejson endpoint)

     ...preview.png?colormap={"1": "#FFFF00", "2":"#FF00FF"} (should be url encoded)

#### Option 2
base64 encoded dict (seems to be a nice option but pretty custom)
    
    ...preview.png?colormap=dsaidhauhdasiojhd928e1uqwijs09qus0x8ahxas

#### Option 3
List defined colormap value (restricted to hex color definition and custom key/color 1,#fff form)

     ...preview.png?colormap=1,FFFF00&colormap=2,FF00FF

I think it's fine to go with **option 1**. It's not the best solution but:
- Starlette will take care of the urlencode/decode part, so we don't really need to worry in the API
- using base64 encoded dict in query string is not something widely used/accepted
- The third option looks more like proper HTTP query string but still custom because we will need to encode `2,FF00FF`

Notes:
I also though about something like 
``` 
colormap=FFFF00&colormap_stop=1,&colormap= FF00FF&colormap_stop=2
``` 

which will 💯 match the specs but still really verbose and not as `easy` as option 1

cc @kylebarron @geospatial-jeff @ingalls

**Pull request merged**